### PR TITLE
feat(qwen3): EAGLE-3 Chain Speculative decoding

### PR DESCRIPTION
**Pull request by scatyf3**
This is an PR for issue #443 

### step1: Chain draft
1. Draft model loading
	1. Model Loader ✅
	2. Memory Allocator ✅
2. Draft model farward pass ✅
	1. Chain(for testing) ✅
3. Target model forward pass: Target feature capture for draft injection  ✅
4. Executor, wire eagle spec decoding to current steps  ✅
5. Reserve memory for both target and draft  ✅
6. Frontend: add EAGLE3 args, wire to existing generation pipeline  ✅

### Change for step 1

Kernel:
1. rope without qk norm `eagle3_rope_kernel`
2. prefill over contiguous KV `single_prefill_nhd_noncausal_cuda`
3. decode over contiguous KV

Model
1. Draft model loader `openinfer-qwen3-4b/src/eagle3/loading.rs`
2. Memory Allocator `openinfer-qwen3-4b/src/eagle3/reservation.rs`
3. Draft Forward pass `openinfer-qwen3-4b/src/eagle3/forward.rs`

correctness check: EAGLE3 python golden gate `eagle3_drafter_golden_gate` from `tools/accuracy/dump_qwen3_4b_eagle3_golden.py`

SD Loop
1. reuse dflash's target feature capture `capture_layer_ids` in `process_all_layers_batch_multi`
2. Draft Executor `openinfer-qwen3-4b/src/executor/eagle3_lane.rs`, maintain draft weight,each request's buffer, state, tmp captured_hidden
3. Target Executor: 
	1. add `execute_eagle3_verify` in `openinfer-qwen3-4b/src/executor.rs`
	2. add eagle3 to  `SpeculativeVerify` and `SpeculativeDraft` in `execute_step_on_lane`
4. For scheduling and sampling, we reuse the dflash util

Correctness check: `openinfer-qwen3-4b/tests/eagle3_speculative_gate.rs`
``
Perf:`openinfer-qwen3-4b/tests/eagle3_speculative_perf.rs` and `openinfer-qwen3-4b/tests/eagle3_gsm8k_perf.rs`

Result

chain sd, k=3, max new=256, rtx5070ti, greedy, 10 prompt from GSM8k

| engine | spec OFF | spec ON   | **speedup** |
| --------- | -------- | --------- | ----------- |
| openinfer | 83.9     | **108.7** | **1.30×**   |
| vLLM 0.21 | 81.9     | 101.1     | 1.23×       |

### Next Step: faithful Tree speculative decoding 

The next step is to implement a faithful version, which we need:
1. Draft Tree decoding
3. Target Tree verify
4. Runtime:  CUDA Graph support for dynamic tree
5. KV Management for Tree

### CODE REVIEW

**Comment by scatyf3** [comment]
@xiaguan @CAICAIIs this adds the chain version of EAGLE-3 and is ready for review — let me know if you'd like any changes before merging.

File: openinfer-kernels/csrc/shared/paged_attention.cu [hunk 1]
```diff
@@ -758,6 +758,135 @@ int single_prefill_nhd_noncausal_cuda_hd64(
   OPENINFER_FFI_GUARD_END(-1)
 }
 
+// Causal variant of single_prefill_nhd_noncausal_cuda: identical NHD token-major
+// layout (q/output [seq, q_dim], k/v [max_seq, kv_dim]) but a causal mask, so the
+// N query tokens at the tail of the cache attend only positions <= their own.
+// Used for EAGLE-3's teacher-forced prefill in one batched forward (query i at
+// absolute position kv_len - seq_len + i, FlashInfer's causal alignment).
+int single_prefill_nhd_causal_cuda(
```
Comment (Ma1oneZhang) [inline openinfer-kernels/csrc/shared/paged_attention.cu:766]: Is it possible to leverage the kernel provided by flashinfer, rather than implementing a custom one from scratch?

File: openinfer-kernels/csrc/shared/prefill_attention.cu [hunk 2]
```diff
@@ -296,4 +359,35 @@ int dflash_qk_norm_rope_cuda(
     return static_cast<int>(cudaGetLastError());
 }
 
+// Plain RoPE (no QK-norm) for EAGLE-3. Launches eagle3_rope_kernel.
+int eagle3_rope_cuda(
```
Comment (Ma1oneZhang) [inline openinfer-kernels/csrc/shared/prefill_attention.cu:363]: Same as the line 211

File: openinfer-kernels/src/ops/attention.rs [hunk 3]
```diff
@@ -858,6 +914,127 @@ pub fn single_prefill_nhd_noncausal_into(
...
 
+/// Single-query **decode** over a contiguous NHD KV cache — the draft chain's
+/// per-step attention. One query (`q.seq_len == 1`) attends the whole `[0, kv_len)`
+/// prefix.
+///
+/// `q`/`output` are `[q_dim, 1]` and the k/v caches are the request's own whole
+/// buffers `[kv_dim, max_seq_len]` (NHD token-major). No RoPE inside — the caller
+/// applies [`eagle3_rope_into`] first.
+#[allow(clippy::too_many_arguments)]
+pub fn single_decode_nhd_into(
```
Comment (Ma1oneZhang) [inline openinfer-kernels/src/ops/attention.rs:925]: Same as the prefill kernel

File: openinfer-qwen3/src/eagle3/forward.rs [hunk 4]
```diff
@@ -0,0 +1,1090 @@
...
+        // 8. Output projection.
+        ops::gemm_into(
+            ctx,
+            &self.midlayer.o_proj,
+            &scratch.attn_out,
+            &mut scratch.o,
+        );
+
+        // 9. Residual + post-attention norm (hidden += o in place; normed_post = norm(hidden)).
+        openinfer_kernels::ops::fused_add_rms_norm_round_batch_into(
```
Comment (Ma1oneZhang) [inline openinfer-qwen3/src/eagle3/forward.rs:275]: Please consider that kernel fusion (with FlashInfer library)

File: openinfer-qwen3/src/eagle3/forward.rs [hunk 5]
```diff
@@ -0,0 +1,1090 @@
...
+
+        state.cached_len = kv_len;
+        Ok(&scratch.logits)
+    }
+
+    /// Teacher-forced prefil for EAGLE Draft
+    /// Returns `(logits [draft_vocab, N], last_hidden [hidden, 1])`
+    /// Buffers are allocated inline (prefill is one-shot, `N` varies). Does not use
+    /// the single-token `Eagle3Scratch`. `tokens[i]` sits at `start_position + i`.
+    pub(crate) fn prefill_batched(
```
Comment (Ma1oneZhang) [inline openinfer-qwen3/src/eagle3/forward.rs:335]: Same as the decode stage(Draft forward)

**Review by Ma1oneZhang — COMMENTED** [review]
First of all, thank you for your contribution.

Second, the pr is large that i can not review it at once. **I’d suggest splitting this large PR into smaller PRs.** One approach would be to first implement a basic Eagle3 inference, and then handle the integration with the overall framework in a follow-up PR.

Third, I’d suggest **we prioritize reusing code from flashinfer’s kernels**, rather than introducing redundant implementations. That would make the changes more maintainable.

Thanks for your contribution again! ^ ^

**Comment by scatyf3** [comment]
@Ma1oneZhang Thank you so much for the thorough review🌹 I really appreciate it. Here's my plan:
1. I'll update the code in this PR based on your suggestions.
2. Once those changes are in, I'll split the work into several smaller PRs and merge them one by one for easier review.

File: openinfer-qwen3/src/eagle3/forward.rs (hunk 4 above)
Comment (scatyf3) [reply openinfer-qwen3/src/eagle3/forward.rs:275]: We use  `fused_add_rms_norm_round_batch_into` kernel is because
1. align with dflash(dflash.rs:706) implementation
2. to keep our implementation bit exactness with huggingface 

for flashinfer, we compute x = input+residual using fp32
```cpp
float x = float(input_vec[j]);
x += float(residual_vec[j]);   
sum_sq += x * x;               
x_vec[j] = x;                  
```

but for huggingface, it use bf16 to perform residual add
```python
hidden = residual + attn_out # bf16
normed = RMSNorm(hidden) # use bf 16 hidden state, than upcast to fp32
```

thus to garentee bit exactness, we use handwritten `fused_add_rms_norm_round_batch_into`, where it computation flow is aligned with huggingface
```c++
T rounded = static_cast<T>(float(hidden_vec[j]) + float(residual_vec[j])); // bf16
float x = float(rounded);      // fp32
sum_sq += x * x;               
rounded_vec[j] = x;

File: openinfer-kernels/csrc/shared/paged_attention.cu (hunk 1 above)
Comment (scatyf3) [reply openinfer-kernels/csrc/shared/paged_attention.cu:766]: for eagle prefill and decode, we already use flashinfer warpper

File: openinfer-kernels/csrc/shared/prefill_attention.cu (hunk 2 above)
Comment (scatyf3) [reply openinfer-kernels/csrc/shared/prefill_attention.cu:363]: for `eagle3_rope_cuda`,  my code is align with dflash's choice, where we both use handwritten rope. a further optimization can change this 2 handwritten kernel to flashinfer version.

**Comment by Ma1oneZhang** [comment]
Thanks for your patience! 

Regarding your comment about bit-wise precision alignment, I fully understand your concern. However, I think a simple Triton implementation with AOT would be better – it has less overhead and is easier to fully understand, and it can be tested in python for better testability.

**Comment by scatyf3** [comment]
This PR is closed because I'll split my change into multiple sub PR for better code review. The first PR is in https://github.com/openinfer-project/openinfer/pull/647
